src/main/scala/com/mozilla/telemetry/learning/federated/FederatedLearningSearchOptimizerS3Sink.scala

```diff
@@ -0,0 +1,151 @@
+        val learningRates: Array[Double] = state.learningRates
+        val previousGradient: Option[Array[Double]] = state.gradient
+
+        val gradient = aggregate.avgUpdates
+        val InternalResult(newWeights, newLearningRates) = FederatedLearningRPropOptimizer.fit(lastWeights, gradient, previousGradient, learningRates)
+
+        val newIteration = iteration + 1
+
+        // val log = Array(OptimisationLogLine(aggregate.window.start, aggregate.modelVersion.toLong, aggregate.avgLoss))
+
+        val newState = OptimisationState(newIteration, newWeights, newLearningRates, Option(gradient))
+        writeModel(ModelOutput(newWeights.map(math.round(_).toInt), newIteration))
+
+        writeState(newState)
+        state = newState
+    }
+  }
+
+  private[federated] def writeModel(modelOutput: ModelOutput): Unit = {
+    implicit val formats = DefaultFormats
+    val jsonModel = Serialization.write(modelOutput)
+
+    val conf = SparkHadoopUtil.get.conf
+    val modelOutputPath = new Path(outputPath)
+
+    val fs = modelOutputPath.getFileSystem(conf)
+    if (!fs.exists(modelOutputPath)) {
+      fs.mkdirs(modelOutputPath)
+    }
+
+
+    val latestStream = fs.create(new Path(outputPath + "/" + "latest.json"))
+    latestStream.writeBytes(jsonModel)
+    latestStream.close()
+
+    val versionedStream = fs.create(new Path(outputPath + "/" + modelOutput.iteration + ".json"))
+    versionedStream.writeBytes(jsonModel)
+    versionedStream.close()
+
+    fs.close()
+  }
+
+  private[federated] def writeState(state: OptimisationState): Unit = {
+    implicit val formats = DefaultFormats
+    val jsonState = Serialization.write(state)
+
+    val fileName = "STATE-" + state.iteration
+    val path = new Path(stateCheckpointPath + "/" + fileName)
+
+    val fs = path.getFileSystem(SparkHadoopUtil.get.conf)
+    val outputStream = fs.create(path)
+    outputStream.writeBytes(jsonState)
+
+    outputStream.close()
+    fs.close()
+  }
+
+  private[federated] def initState(): OptimisationState = {
+    val conf = SparkHadoopUtil.get.conf
+    val checkpointPath = new Path(stateCheckpointPath)
+    val fs = checkpointPath.getFileSystem(conf)
+    if (!fs.exists(checkpointPath)) {
+      fs.mkdirs(checkpointPath)
+    }
+
+    val statuses = fs.listStatus(checkpointPath)
+    val latestFileOpt = if (statuses != null) {
+      val paths = statuses.map(_.getPath)
+      val sorted = paths.sortWith { case (p1, p2) => p1.getName > p2.getName }
+      sorted.headOption
+    } else {
+      None
+    }
+    try {
+      latestFileOpt match {
+        case Some(latestFile) =>
+          val fis = fs.open(latestFile)
+          val fileContents: String = IOUtils.toString(fis)
+          fis.close()
+
+          implicit val formats = DefaultFormats
+          Serialization.read[OptimisationState](fileContents)
+        case None =>
+          OptimisationState(0, StartingWeights, Array.fill(NumberOfFeatures)(StartingLearningRate), None)
```

Comment:
Can we add a config option that points to a file we can read in for bootstrapping OptimisationState? I think that might come in really handy if we need to restart from a particular point

Reply:
Good idea, adding this option.